# <u>**Part 11**

In [7]:
import pandas as pd
import numpy as np
import json

In [2]:
data = {
    'size': ['XL', 'L', 'M', 'L', 'M'],
    'color': ['red', 'green', 'blue', 'green', 'red'],
    'gender': ['female', 'male', 'male', 'female', 'female'],
    'price': [199.0, 89.0, 99.0, 129.0, 79.0],
    'weight': [500, 450, 300, 380, 410],
    'bought': ['yes', 'no', 'yes', 'no', 'yes']
}

df = pd.DataFrame(data)
for col in ['size', 'color', 'gender', 'bought']:
    df[col] = df[col].astype('category')

df['weight'] = df['weight'].astype('float')

In [6]:
market = {
    "profile":{
        "0":"^AEX (Holandia)",
        "1":"^ATX (Austria)",
        "2":"^ATXC (Grecja)",
        "3":"^BEL20 (Belgia)",
        "4":"^BUX (W\u0119gry)",
        "5":"^CAC40 (Francja)",
        "6":"^DAX (Niemcy)",
        "7":"^FTSE100 (Wielka Brytania)",
        "8":"^FTSE250 (Wielka Brytania)",
        "9":"^FTSEMIB (W\u0142ochy)",
        "10":"^IBEX35 (Hiszpania)",
        "11":"^OMXC20 (Dania)",
        "12":"^OMXH25 (Finlandia)",
        "13":"^OMXS30 (Szwecja)",
        "14":"^OSEAX (Norwegia)",
        "15":"^PSI20 (Portugalia)",
        "16":"^PX50 (Czechy)",
        "17":"^RDXEUR (Rosja)",
        "18":"^RTS (Rosja)",
        "19":"^SMI (Szwajcaria)",
        "20":"^XU100 (Turcja)",
        "21":"WIG20 (Polska)"
    },
    "time":{
        "0":"14:12",
        "1":"14:11",
        "2":"14:13",
        "3":"14:11",
        "4":"14:12",
        "5":"14:12",
        "6":"14:14",
        "7":"14:12",
        "8":"14:12",
        "9":"14:12",
        "10":"14:12",
        "11":"14:27",
        "12":"14:26",
        "13":"14:27",
        "14":"14:12",
        "15":"14:13",
        "16":"14:12",
        "17":"14:11",
        "18":"14:11",
        "19":"14:11",
        "20":"14:12",
        "21":"14:29"
    },
    "price":{
        "0":548.73,
        "1":2147.9,
        "2":621.82,
        "3":3251.39,
        "4":32831.01,
        "5":4839.46,
        "6":12758.52,
        "7":5945.44,
        "8":17038.88,
        "9":19080.16,
        "10":6755.8,
        "11":1336.57,
        "12":10037.06,
        "13":1802.43,
        "14":739.99,
        "15":4119.97,
        "16":870.57,
        "17":1264.48,
        "18":1524.18,
        "19":10371.42,
        "20":1104.41,
        "21":1695.96
    },
    "change":{
        "0":7.95,
        "1":17.68,
        "2":-2.8,
        "3":30.51,
        "4":534.27,
        "5":66.62,
        "6":164.13,
        "7":115.98,
        "8":217.22,
        "9":184.57,
        "10":107.2,
        "11":0.58,
        "12":21.87,
        "13":20.51,
        "14":1.36,
        "15":-1.97,
        "16":2.52,
        "17":2.21,
        "18":-2.09,
        "19":15.85,
        "20":8.25,
        "21":2.87
    },
    "pct_change":{
        "0":"(+1.47%)",
        "1":"(+0.83%)",
        "2":"(-0.45%)",
        "3":"(+0.95%)",
        "4":"(+1.65%)",
        "5":"(+1.40%)",
        "6":"(+1.30%)",
        "7":"(+1.99%)",
        "8":"(+1.29%)",
        "9":"(+0.98%)",
        "10":"(+1.61%)",
        "11":"(+0.04%)",
        "12":"(+0.22%)",
        "13":"(+1.15%)",
        "14":"(+0.18%)",
        "15":"(-0.05%)",
        "16":"(+0.29%)",
        "17":"(+0.18%)",
        "18":"(-0.14%)",
        "19":"(+0.15%)",
        "20":"(+0.75%)",
        "21":"(+0.17%)"
    },
    "reference_price":{
        "0":540.78,
        "1":2130.22,
        "2":624.62,
        "3":3220.88,
        "4":32296.74,
        "5":4772.84,
        "6":12594.39,
        "7":5829.46,
        "8":16821.66,
        "9":18895.59,
        "10":6648.6,
        "11":1335.99,
        "12":10015.19,
        "13":1781.92,
        "14":738.63,
        "15":4121.94,
        "16":868.05,
        "17":1262.27,
        "18":1526.27,
        "19":10355.57,
        "20":1096.16,
        "21":1693.09
    },
    "open":{
        "0":546.55,
        "1":2131.47,
        "2":627.48,
        "3":3252.21,
        "4":32421.05,
        "5":4824.17,
        "6":12737.32,
        "7":5829.46,
        "8":16821.66,
        "9":19080.18,
        "10":6717.8,
        "11":1338.82,
        "12":10052.81,
        "13":1791.67,
        "14":738.46,
        "15":4141.19,
        "16":868.53,
        "17":1263.78,
        "18":1526.27,
        "19":10368.65,
        "20":1101.04,
        "21":1701.26
    },
    "low":{
        "0":544.72,
        "1":2118.71,
        "2":621.82,
        "3":3228.77,
        "4":32421.05,
        "5":4803.91,
        "6":12688.33,
        "7":5829.46,
        "8":16821.66,
        "9":18944.29,
        "10":6681.6,
        "11":1322.42,
        "12":10024.27,
        "13":1789.7,
        "14":738.03,
        "15":4119.97,
        "16":862.45,
        "17":1259.57,
        "18":1521.7,
        "19":10334.54,
        "20":1098.82,
        "21":1695.36
    },
    "high":{
        "0":550.72,
        "1":2163.92,
        "2":631.7,
        "3":3266.11,
        "4":32865.43,
        "5":4871.45,
        "6":12830.7,
        "7":5974.88,
        "8":17123.31,
        "9":19231.99,
        "10":6781.1,
        "11":1338.82,
        "12":10107.86,
        "13":1814.53,
        "14":744.44,
        "15":4156.11,
        "16":873.37,
        "17":1274.42,
        "18":1541.51,
        "19":10439.74,
        "20":1106.06,
        "21":1712.27
    }
}

In [8]:
market_json = json.dumps(market, indent = 4) 
print(market_json)

{
    "profile": {
        "0": "^AEX (Holandia)",
        "1": "^ATX (Austria)",
        "2": "^ATXC (Grecja)",
        "3": "^BEL20 (Belgia)",
        "4": "^BUX (W\u0119gry)",
        "5": "^CAC40 (Francja)",
        "6": "^DAX (Niemcy)",
        "7": "^FTSE100 (Wielka Brytania)",
        "8": "^FTSE250 (Wielka Brytania)",
        "9": "^FTSEMIB (W\u0142ochy)",
        "10": "^IBEX35 (Hiszpania)",
        "11": "^OMXC20 (Dania)",
        "12": "^OMXH25 (Finlandia)",
        "13": "^OMXS30 (Szwecja)",
        "14": "^OSEAX (Norwegia)",
        "15": "^PSI20 (Portugalia)",
        "16": "^PX50 (Czechy)",
        "17": "^RDXEUR (Rosja)",
        "18": "^RTS (Rosja)",
        "19": "^SMI (Szwajcaria)",
        "20": "^XU100 (Turcja)",
        "21": "WIG20 (Polska)"
    },
    "time": {
        "0": "14:12",
        "1": "14:11",
        "2": "14:13",
        "3": "14:11",
        "4": "14:12",
        "5": "14:12",
        "6": "14:14",
        "7": "14:12",
        "8": "14:12",
      

### **Exercise 101**

Transform all categorical variables to 0-1 encoding as shown below and print result to the console.



In [3]:
df.head()

size  color  gender  price  weight bought
0   XL    red  female  199.0   500.0    yes
1    L  green    male   89.0   450.0     no
2    M   blue    male   99.0   300.0    yes
3    L  green  female  129.0   380.0     no
4    M    red  female   79.0   410.0    yes

In [4]:
pd.get_dummies(df)

price  weight  size_L  size_M  size_XL  color_blue  color_green  color_red  \
0  199.0   500.0       0       0        1           0            0          1   
1   89.0   450.0       1       0        0           0            1          0   
2   99.0   300.0       0       1        0           1            0          0   
3  129.0   380.0       1       0        0           0            1          0   
4   79.0   410.0       0       1        0           0            0          1   

   gender_female  gender_male  bought_no  bought_yes  
0              1            0          0           1  
1              0            1          1           0  
2              0            1          0           1  
3              1            0          1           0  
4              1            0          0           1

### **Exercise 102**

Once again transform all categorical variables to 0-1 encoding as shown below and print result to the console. This time set drop_first=True


In [5]:
pd.get_dummies(df, drop_first=True)

price  weight  size_M  size_XL  color_green  color_red  gender_male  \
0  199.0   500.0       0        1            0          1            0   
1   89.0   450.0       0        0            1          0            1   
2   99.0   300.0       1        0            0          0            1   
3  129.0   380.0       0        0            1          0            0   
4   79.0   410.0       1        0            0          1            0   

   bought_yes  
0           1  
1           0  
2           1  
3           0  
4           1

### **Exercise 103**

Load the market_json file into a pandas dataframe and assign it to the df variable

In [12]:
df = pd.read_json(market_json)
df

profile   time     price  change pct_change  \
0              ^AEX (Holandia)  14:12    548.73    7.95   (+1.47%)   
1               ^ATX (Austria)  14:11   2147.90   17.68   (+0.83%)   
2               ^ATXC (Grecja)  14:13    621.82   -2.80   (-0.45%)   
3              ^BEL20 (Belgia)  14:11   3251.39   30.51   (+0.95%)   
4                 ^BUX (Węgry)  14:12  32831.01  534.27   (+1.65%)   
5             ^CAC40 (Francja)  14:12   4839.46   66.62   (+1.40%)   
6                ^DAX (Niemcy)  14:14  12758.52  164.13   (+1.30%)   
7   ^FTSE100 (Wielka Brytania)  14:12   5945.44  115.98   (+1.99%)   
8   ^FTSE250 (Wielka Brytania)  14:12  17038.88  217.22   (+1.29%)   
9            ^FTSEMIB (Włochy)  14:12  19080.16  184.57   (+0.98%)   
10         ^IBEX35 (Hiszpania)  14:12   6755.80  107.20   (+1.61%)   
11             ^OMXC20 (Dania)  14:27   1336.57    0.58   (+0.04%)   
12         ^OMXH25 (Finlandia)  14:26  10037.06   21.87   (+0.22%)   
13           ^OMXS30 (Szwecja)  14:27   1802.43   20.51   (+1.15%)   
14           ^OSEAX (Norwegia)  14:12    739.99    1.36   (+0.18%)   
15         ^PSI20 (Portugalia)  14:13   4119.97   -1.97   (-0.05%)   
16              ^PX50 (Czechy)  14:12    870.57    2.52   (+0.29%)   
17             ^RDXEUR (Rosja)  14:11   1264.48    2.21   (+0.18%)   
18                ^RTS (Rosja)  14:11   1524.18   -2.09   (-0.14%)   
19           ^SMI (Szwajcaria)  14:11  10371.42   15.85   (+0.15%)   
20             ^XU100 (Turcja)  14:12   1104.41    8.25   (+0.75%)   
21              WIG20 (Polska)  14:29   1695.96    2.87   (+0.17%)   

    reference_price      open       low      high  
0            540.78    546.55    544.72    550.72  
1           2130.22   2131.47   2118.71   2163.92  
2            624.62    627.48    621.82    631.70  
3           3220.88   3252.21   3228.77   3266.11  
4          32296.74  32421.05  32421.05  32865.43  
5           4772.84   4824.17   4803.91   4871.45  
6          12594.39  12737.32  12688.33  12830.70  
7           5829.46   5829.46   5829.46   5974.88  
8          16821.66  16821.66  16821.66  17123.31  
9          18895.59  19080.18  18944.29  19231.99  
10          6648.60   6717.80   6681.60   6781.10  
11          1335.99   1338.82   1322.42   1338.82  
12         10015.19  10052.81  10024.27  10107.86  
13          1781.92   1791.67   1789.70   1814.53  
14           738.63    738.46    738.03    744.44  
15          4121.94   4141.19   4119.97   4156.11  
16           868.05    868.53    862.45    873.37  
17          1262.27   1263.78   1259.57   1274.42  
18          1526.27   1526.27   1521.70   1541.51  
19         10355.57  10368.65  10334.54  10439.74  
20          1096.16   1101.04   1098.82   1106.06  
21          1693.09   1701.26   1695.36   1712.27

### **Exercise 104**

Extract the country name from the first column and assign to 'country'. 

In [13]:
df['country'] = df['profile'].str.split(n=1, expand=True)[1]
df['country'] = df['country'].apply(lambda i: i[1:-1])
df

profile   time     price  change pct_change  \
0              ^AEX (Holandia)  14:12    548.73    7.95   (+1.47%)   
1               ^ATX (Austria)  14:11   2147.90   17.68   (+0.83%)   
2               ^ATXC (Grecja)  14:13    621.82   -2.80   (-0.45%)   
3              ^BEL20 (Belgia)  14:11   3251.39   30.51   (+0.95%)   
4                 ^BUX (Węgry)  14:12  32831.01  534.27   (+1.65%)   
5             ^CAC40 (Francja)  14:12   4839.46   66.62   (+1.40%)   
6                ^DAX (Niemcy)  14:14  12758.52  164.13   (+1.30%)   
7   ^FTSE100 (Wielka Brytania)  14:12   5945.44  115.98   (+1.99%)   
8   ^FTSE250 (Wielka Brytania)  14:12  17038.88  217.22   (+1.29%)   
9            ^FTSEMIB (Włochy)  14:12  19080.16  184.57   (+0.98%)   
10         ^IBEX35 (Hiszpania)  14:12   6755.80  107.20   (+1.61%)   
11             ^OMXC20 (Dania)  14:27   1336.57    0.58   (+0.04%)   
12         ^OMXH25 (Finlandia)  14:26  10037.06   21.87   (+0.22%)   
13           ^OMXS30 (Szwecja)  14:27   1802.43   20.51   (+1.15%)   
14           ^OSEAX (Norwegia)  14:12    739.99    1.36   (+0.18%)   
15         ^PSI20 (Portugalia)  14:13   4119.97   -1.97   (-0.05%)   
16              ^PX50 (Czechy)  14:12    870.57    2.52   (+0.29%)   
17             ^RDXEUR (Rosja)  14:11   1264.48    2.21   (+0.18%)   
18                ^RTS (Rosja)  14:11   1524.18   -2.09   (-0.14%)   
19           ^SMI (Szwajcaria)  14:11  10371.42   15.85   (+0.15%)   
20             ^XU100 (Turcja)  14:12   1104.41    8.25   (+0.75%)   
21              WIG20 (Polska)  14:29   1695.96    2.87   (+0.17%)   

    reference_price      open       low      high          country  
0            540.78    546.55    544.72    550.72         Holandia  
1           2130.22   2131.47   2118.71   2163.92          Austria  
2            624.62    627.48    621.82    631.70           Grecja  
3           3220.88   3252.21   3228.77   3266.11           Belgia  
4          32296.74  32421.05  32421.05  32865.43            Węgry  
5           4772.84   4824.17   4803.91   4871.45          Francja  
6          12594.39  12737.32  12688.33  12830.70           Niemcy  
7           5829.46   5829.46   5829.46   5974.88  Wielka Brytania  
8          16821.66  16821.66  16821.66  17123.31  Wielka Brytania  
9          18895.59  19080.18  18944.29  19231.99           Włochy  
10          6648.60   6717.80   6681.60   6781.10        Hiszpania  
11          1335.99   1338.82   1322.42   1338.82            Dania  
12         10015.19  10052.81  10024.27  10107.86        Finlandia  
13          1781.92   1791.67   1789.70   1814.53          Szwecja  
14           738.63    738.46    738.03    744.44         Norwegia  
15          4121.94   4141.19   4119.97   4156.11       Portugalia  
16           868.05    868.53    862.45    873.37           Czechy  
17          1262.27   1263.78   1259.57   1274.42            Rosja  
18          1526.27   1526.27   1521.70   1541.51            Rosja  
19         10355.57  10368.65  10334.54  10439.74       Szwajcaria  
20          1096.16   1101.04   1098.82   1106.06           Turcja  
21          1693.09   1701.26   1695.36   1712.27           Polska

### **Exercise 105**

Extract country names (sorted alphabetically) that occur more than once as a list and print it to the console.

In [22]:
sorted(df['country'].value_counts()[df['country'].value_counts() > 1].index)

['Rosja', 'Wielka Brytania']

### **Exercise 106**

The pct_change column has the object type. Convert this column to float type.



In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22 entries, 0 to 21
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   profile          22 non-null     object 
 1   time             22 non-null     object 
 2   price            22 non-null     float64
 3   change           22 non-null     float64
 4   pct_change       22 non-null     object 
 5   reference_price  22 non-null     float64
 6   open             22 non-null     float64
 7   low              22 non-null     float64
 8   high             22 non-null     float64
 9   country          22 non-null     object 
dtypes: float64(6), object(4)
memory usage: 1.9+ KB


In [24]:
df['pct_change'] = df['pct_change'].apply(lambda item: float(item[1:-2]))
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22 entries, 0 to 21
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   profile          22 non-null     object 
 1   time             22 non-null     object 
 2   price            22 non-null     float64
 3   change           22 non-null     float64
 4   pct_change       22 non-null     float64
 5   reference_price  22 non-null     float64
 6   open             22 non-null     float64
 7   low              22 non-null     float64
 8   high             22 non-null     float64
 9   country          22 non-null     object 
dtypes: float64(7), object(3)
memory usage: 1.9+ KB


In [25]:
df.head(5)

profile   time     price  change  pct_change  reference_price  \
0  ^AEX (Holandia)  14:12    548.73    7.95        1.47           540.78   
1   ^ATX (Austria)  14:11   2147.90   17.68        0.83          2130.22   
2   ^ATXC (Grecja)  14:13    621.82   -2.80       -0.45           624.62   
3  ^BEL20 (Belgia)  14:11   3251.39   30.51        0.95          3220.88   
4     ^BUX (Węgry)  14:12  32831.01  534.27        1.65         32296.74   

       open       low      high   country  
0    546.55    544.72    550.72  Holandia  
1   2131.47   2118.71   2163.92   Austria  
2    627.48    621.82    631.70    Grecja  
3   3252.21   3228.77   3266.11    Belgia  
4  32421.05  32421.05  32865.43     Węgry

### **Exercise 107**

Perform the following steps

* print to the console row with the highest *pct_change*
* print a blank line
* print to the console row loest *pct_change*

In [26]:
print(df.iloc[df['pct_change'].argmax()])
print()
print(df.iloc[df['pct_change'].argmin()])

profile            ^FTSE100 (Wielka Brytania)
time                                    14:12
price                                 5945.44
change                                 115.98
pct_change                               1.99
reference_price                       5829.46
open                                  5829.46
low                                   5829.46
high                                  5974.88
country                       Wielka Brytania
Name: 7, dtype: object

profile            ^ATXC (Grecja)
time                        14:13
price                      621.82
change                       -2.8
pct_change                  -0.45
reference_price            624.62
open                       627.48
low                        621.82
high                        631.7
country                    Grecja
Name: 2, dtype: object


### **Exercise 108**

Extract all rows of this object for which the pct_change column is greater than or equal to zero and save to 'positive.json' file. Do the same for the negative values and save to the 'negative.json' file.

In [30]:
df[df['pct_change'] >= 0].to_json('positive.json', indent=4)
df[df['pct_change'] < 0].to_json('negative.json', indent=4)

### **Exercise 109**

Load amazon and tesla stock price information into two different dataframes

In [31]:
import pandas_datareader as web

In [33]:
amzn = web.DataReader('AMZN', data_source='stooq', start='2020-01-01', end='2022-01-01')
amzn

Open      High       Low     Close    Volume
Date                                                       
2021-12-31  168.956  169.3500  166.5580  166.7170  47830380
2021-12-30  169.700  170.8880  168.5240  168.6440  37583100
2021-12-29  170.840  171.2120  168.6010  169.2010  35754040
2021-12-28  170.182  172.1760  169.1360  170.6610  54711860
2021-12-27  171.037  172.9430  169.2160  169.6700  58846820
...             ...       ...       ...       ...       ...
2020-01-08   94.902   95.5500   94.3220   94.5985  70239320
2020-01-07   95.225   95.6945   94.6020   95.3430  82680200
2020-01-06   93.000   95.1845   93.0000   95.1440  81313960
2020-01-03   93.225   94.3100   93.2250   93.7485  75332080
2020-01-02   93.750   94.9005   93.2075   94.9005  80718200

[505 rows x 5 columns]

In [34]:
tsla = web.DataReader('TSLA', data_source='stooq', start='2020-01-01', end='2022-01-01')
tsla

Open      High       Low     Close     Volume
Date                                                         
2021-12-31  357.8150  360.6670  351.5300  352.2600   40733625
2021-12-30  353.7770  365.1830  351.0500  356.7800   47040939
2021-12-29  366.2130  368.0000  354.7130  362.0630   56154045
2021-12-28  369.8300  373.0000  359.4730  362.8230   60323907
2021-12-27  357.8900  372.3330  356.9050  364.6470   71145819
...              ...       ...       ...       ...        ...
2020-01-08   31.5800   33.2333   31.2153   32.8093  467990895
2020-01-07   30.7600   31.4420   30.2237   31.2707  273137070
2020-01-06   29.3647   30.1040   29.3333   30.1027  152362485
2020-01-03   29.3667   30.2667   29.1280   29.5340  266920455
2020-01-02   28.3000   28.7130   28.1140   28.6840  143375790

[505 rows x 5 columns]

### **Exercise 110**

Replace the column names in these objects by adding prefix 'amzn' to each column in the amazon DataFrame and prefix 'tsla' to each column in the tesla DataFrame. Also convert uppercase letters to lowercase

In [35]:
amzn.columns = ['amzn_' + col.lower() for col in amzn.columns]
tsla.columns = ['tsla_' + col.lower() for col in tsla.columns]

In [36]:
amzn.head(5)

amzn_open  amzn_high  amzn_low  amzn_close  amzn_volume
Date                                                               
2021-12-31    168.956    169.350   166.558     166.717     47830380
2021-12-30    169.700    170.888   168.524     168.644     37583100
2021-12-29    170.840    171.212   168.601     169.201     35754040
2021-12-28    170.182    172.176   169.136     170.661     54711860
2021-12-27    171.037    172.943   169.216     169.670     58846820

In [37]:
tsla.head(5)

tsla_open  tsla_high  tsla_low  tsla_close  tsla_volume
Date                                                               
2021-12-31    357.815    360.667   351.530     352.260     40733625
2021-12-30    353.777    365.183   351.050     356.780     47040939
2021-12-29    366.213    368.000   354.713     362.063     56154045
2021-12-28    369.830    373.000   359.473     362.823     60323907
2021-12-27    357.890    372.333   356.905     364.647     71145819